File: day11/py/puzzle2.py
#!/usr/bin/env python

from util import load_monkeys
from functools import reduce

def solve():
    with open('../input.txt', 'r') as f:
        monkeys = load_monkeys(f)   

    limit = reduce(lambda a, b: a * b, [m.test_value for m in monkeys])

    for _round in range(0, 10_000):
        for monkey in monkeys:
            if monkey.inspect():
                monkey.items = [i % limit for i in monkey.items]
                monkey.act(monkeys)

    sorted_inspections = sorted([m.inspections for m in monkeys])

    return sorted_inspections[-2] * sorted_inspections[-1]

if __name__ == "__main__":
    solution = solve()
    print(solution)

File: day11/cxx/puzzle2.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <numeric>
#include <vector>

#include "range.hpp"
#include "util.hpp"

using namespace std;

const auto solve()
{
    ifstream f{"../input.txt"};

    auto monkeys = load_monkeys(f);

    auto limit = std::transform_reduce(
        monkeys.cbegin(),
        monkeys.cend(),
        1ull,
        std::multiplies{},
        [](const Monkey& m) { return m.test_value; }
    );

    for (auto _round: iter_range(0, 10'000))
    {
        for (auto& monkey: monkeys)
        {
            if (monkey.inspect())
            {
                std::transform(
                    monkey.items.cbegin(),
                    monkey.items.cend(),
                    monkey.items.begin(),
                    [=](const uint64_t v) { return v % limit; } 
                );

                monkey.act(monkeys);
            }
        }
    }


    vector<uint64_t> inspections(monkeys.size());

    std::transform(
        monkeys.cbegin(),
        monkeys.cend(),
        inspections.begin(),
        [](const Monkey& monkey) { return monkey.inspections; }
    );

    std::sort(inspections.begin(), inspections.end());

    auto v1 = *(inspections.rbegin());
    auto v2 = *(inspections.rbegin() + 1);

    return v1 * v2;    
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}